docs: First shot at run in prod. (#1168)
cosmos/sdk-tutorials: academy/xl-run-prod/4-network.md

# Prepare and connect to other nodes

With the genesis created and received, a node operator needs to join the eventual network. In practice this means two things:

1. To open your node to connections from other.
2. To know where are the other nodes, or at least a subset of them so that your node can attempt to connect to them.

In this section, you concern yourself with Tendermint and the peer-to-peer network. Other niceties like gRPC and REST into your Cosmos application are different concerns.

## Set up 

As a node operator, from the time of genesis or at any time in the future, and on each machine, you first run an `init` command to at least set up the folders and pick an ASCII-only moniker:

```sh
$ ./myprojectd init lascaux
```

Overwrite the genesis created again with the real agreed one. While you are at it, you can make it read-only:

```sh
$ curl http://example.com/genesis.json -o ~/.myprojectd/config/genesis.json
$ chmod a-wx ~/.myprojectd/config/genesis.json
```

The `init` command has also created a number of configuration files:

```sh
$ ls ~/.myprojectd/config
```

Should return:

```txt
addrbook.json
app.toml
client.toml

> config.toml <-- configuration for Tendermint

Reviewer: Good idea to explain which file is used for what. Maybe we should add comments for all?

genesis.json
gentx
node_key.json
priv_validator_key.json
```

## Open your node

In the [`config.toml` file](https://docs.tendermint.com/master/nodes/configuration.html), you can configure the open ports. The important piece is:

```toml
# Address to listen for incoming connections
laddr = "tcp://0.0.0.0:26656"
```

> Here it listens on port `26656` of all IP addresses. Define or find out your publicly accessible IP address, for instance `74.6.231.20`.

Reviewer: ```suggestion
Here it listens on port `26656` of all IP addresses. Define or find out your publicly accessible IP address, for instance `172.217.22.14`.
```

If we are using a public IP address, let's use a known one instead of a random one :)

If you use a DNS-resolvable name like `lascaux.myproject.example.com`, you can use that as well instead of the IP address. Keep in mind that a name is subject to the DNS being well configured and working well.

Add it too so that, whenever your node contacts a new node, yours can tell it which address is preferred:

```toml

> external_address = "74.6.231.20:26656"

Reviewer: ```suggestion
external_address = "172.217.22.14:26656" # replace by your own
```

```

The other piece of information that uniquely identifies your node is your node key. Its private key is stored in `~/.myprojectd/config/node_key.json`. The public key is that by which your peers will know your node. You can compute the public key with the Tendermint command:

```sh
$ ./myprojectd tendermint show-node-id
```

This should return something like:

```txt
ce1c54ea7a2c50b4b9f2f869faf8fa4d1a1cf43a
```

If you lose `node_key.json`, or have it stolen, it is not as serious as if you lost your token's private key. Your node can always recreate it and let your peers know about the new key, no hard feelings. Its location is mentioned in `config.toml` on the line `node_key_file = "config/node_key.json"`.

The node key also exists so that your own node can identify itself if it tried to connect to itself via a circuitous peer-to-peer route, and therefore ought to cut the useless connection.

In short here is the information you need to share with other early participants in the network:

> * Listen address, for instance: `"tcp://74.6.231.20:26656"`.

Reviewer: ```suggestion
* Listen address, for instance: `"tcp://172.217.22.14:26656"`.
```

* Node key, for instance: `ce1c54ea7a2c50b4b9f2f869faf8fa4d1a1cf43a`.

The shorthand for this information is written and exchanged as such:

```txt

> ce1c54ea7a2c50b4b9f2f869faf8fa4d1a1cf43a@74.6.231.20:26656

Reviewer: ```suggestion
ce1c54ea7a2c50b4b9f2f869faf8fa4d1a1cf43a@172.217.22.14:26656
```

5735836cbaa747e013e47b11839db2c2990b918a@121.37.49.12:26656"
```

If one of the operators informs you that they node behaves as a seed node, then you add it under:

<CodeGroup>

<CodeGroupItem title="Up to v0.34" active>

```toml
seeds = "432d816d0a1648c5bc3f060bd28dea6ff13cb413@216.58.206.174:26656"
```

</CodeGroupItem>

<CodeGroupItem title="From v0.35">

```toml
bootstrap_peers = "432d816d0a1648c5bc3f060bd28dea6ff13cb413@216.58.206.174:26656"
```

</CodeGroupItem>

</CodeGroup>

You can also take this opportunity to document the list of peers on your _production_ repository (the same that hosts the genesis file).

Note that you are not obliged to put all the known peers in your `persistent_peers`. You may well choose to put them only those you trust.

## Further network configuration

Setting up your peer and identifying other peers is important. However this is not the only network configuration available. Look into `~/.myprojectd/config/config.toml` for tweaks.

If you change parameters in this file, you are not going to affect the ability of the network to reach consensus on blocks. Parameters that are necessary for consensus are all in the genesis file.

> [Parameters in `config.toml`](https://docs.tendermint.com/master/nodes/configuration.html) can be divided into two broad categories:

Reviewer: Tendermint advice to not put any link to `/master` but pin the version. The `/master` endpoint does not exist anymore.

```suggestion
[Parameters in `config.toml`](https://docs.tendermint.com/v0.34/tendermint-core/configuration.html) can be divided into two broad categories:
```

> 1. Network scoped. By changing them you change the posture of your node at the risk of disrupting the ability of other nodes to communicate with yours. Examples include `max_num_inbound_peers` (up to v0.34), `max_connections` (from v0.35), and `handshake_timeout`.

Reviewer: ```suggestion
1. Network scoped. By changing them you change the posture of your node at the risk of disrupting the ability of other nodes to communicate with yours. Examples include `max_num_inbound_peers, and `handshake_timeout`.
```

Tendermint v0.35 has been discontinued, we should not mention it.

Author: Wait, this means that we need to keep `max_connections` or `max_num_inbound_peers`?

Reviewer: We should keep `max_num_inbound_peers`
